### PR TITLE
Add test cases for the custom mutator

### PR DESCRIPTION
**Pull request by h1994st**
Ready for review...

- Update the Makefile in examples/custom_mutators
- Add a test program for testing the custom mutator
- Update test.sh for testing the custom mutator
- Update the result checking criteria of the custom mutator in test.sh
- Fix memory reallocation bug in "afl-fuzz-one.c:1624"

### CODE REVIEW

**Comment by domenukk** [comment]
`Write_to_testcase` was a wip but got removed again 🙃
It would be good to unify the api similar to the new `pre_save` https://github.com/AFLplusplus/AFLplusplus/issues/273#issuecomment-604580090

File: src/afl-fuzz-mutators.c [hunk 1]
```diff
@@ -192,7 +192,8 @@ void load_custom_mutator(afl_state_t *afl, const char *fn) {
 
   /* Initialize the custom mutator */
   if (afl->mutator->afl_custom_init)
-    afl->mutator->afl_custom_init(afl, rand_below(afl, 0xFFFFFFFF));
+    afl->mutator->data = afl->mutator->afl_custom_init(
+      afl, rand_below(afl, 0xFFFFFFFF));
```
Comment (domenukk) [inline src/afl-fuzz-mutators.c]: Oh great find!

Comment (h1994st) [reply src/afl-fuzz-mutators.c]: Also found some bugs in the custom mutator 😂
Try to solve them

**Comment by vanhauser-thc** [comment]
build fails with
```
[-] PROGRAM ABORT : Trimmed data returned by custom mutator is larger than original data

         Location : trim_case_custom(), src/afl-fuzz-mutators.c:237
```

**Comment by h1994st** [comment]
> build fails with
> 
> ```
> [-] PROGRAM ABORT : Trimmed data returned by custom mutator is larger than original data
> 
>          Location : trim_case_custom(), src/afl-fuzz-mutators.c:237
> ```

Yes, there are some memory bugs in `afl-fuzz` and `example.c` ...
I am still inspecting them 😂

**Comment by h1994st** [comment]
There is a weird memory bug.

For my fork branch, while running `make tests`, `afl-fuzz` will throw an error: `realloc(): invalid next size`. (Although `test.sh` says the Python mutator works correctly, but it will trigger the same error as the C mutator if it runs a longer time.)

Here is the backtrace:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff7a24801 in __GI_abort () at abort.c:79
#2  0x00007ffff7a6d897 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7b9ab9a "%s\n") at ../sysdeps/posix/libc_fatal.c:181
#3  0x00007ffff7a7490a in malloc_printerr (str=str@entry=0x7ffff7b98ed3 "realloc(): invalid next size") at malloc.c:5350
#4  0x00007ffff7a799b4 in _int_realloc (av=av@entry=0x7ffff7dcfc40 <main_arena>, oldp=oldp@entry=0x5555557c4dd0, oldsize=oldsize@entry=80, nb=nb@entry=144) at malloc.c:4534
#5  0x00007ffff7a7ccfd in __GI___libc_realloc (oldmem=0x5555557c4de0, bytes=128) at malloc.c:3221
#6  0x000055555555da3e in DFL_ck_realloc (size=128, orig=<optimized out>) at include/alloc-inl.h:144
#7  ck_maybe_grow (buf=0x7ffff7fbc560, size=0x7ffff7fbc568, size_needed=<optimized out>) at include/alloc-inl.h:813
#8  0x00005555555749d6 in fuzz_one_original (afl=0x7ffff7eca010) at src/afl-fuzz-one.c:1979
#9  0x0000555555578395 in fuzz_one (afl=0x7ffff7eca010) at src/afl-fuzz-one.c:4403
#10 0x000055555555b47f in main (argc=9, argv_orig=<optimized out>, envp=<optimized out>) at src/afl-fuzz.c:1085
```

Any comments will be appreciated! I will double-check it tomorrow.

**Comment by domenukk** [comment]
Please try if a `make clean` && `make` may fix this issue. This error message no longer exists in `dev` so you probably didn't build the version where @hexcoder- fixed some of my initial bugs.

**Comment by vanhauser-thc** [comment]
Have to tried valgrind? Or compiling everything incl the custom lib with asan?

**Comment by h1994st** [comment]
Finally solved the bug ...😂

The deleted line in "src/afl-fuzz-one.c:1624" may modify `afl->out_size`. However, the corresponding `afl->out_buf` is not changed. Thus, the future memory reallocation invoked by `ck_maybe_grow ` may lead to inconsistent buffer and size. Any memory operations on the wrong buffer may modify heap metadata and result in the reallocation bug/heap over flow (i.e., `realloc(): invalid next size`).

**Comment by h1994st** [comment]
> `Write_to_testcase` was a wip but got removed again 🙃
> It would be good to unify the api similar to the new `pre_save` [#273 (comment)](https://github.com/AFLplusplus/AFLplusplus/issues/273#issuecomment-604580090)

The test cases work correctly now. I may refactor the custom mutator APIs later...

**Comment by domenukk** [comment]
I've cleaned up some things and created wip/mutators.
On my machine, the test cases still time out:
```bash
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...
[*] Entering queue cycle 40.
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...
[*] Entering queue cycle 41.
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...
[*] Entering queue cycle 42.
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...
[*] Entering queue cycle 43.
[*] Fuzzing test case #0 (1 total, 0 uniq crashes found)...


+++ Testing aborted programmatically +++
[!] Time limit was reached
[+] We're done here. Have a nice day!

CUT------------------------------------------------------------------CUT
[!] afl-fuzz is not working correctly with the C mutator
```

**Comment by domenukk** [comment]
My bad. Fixed :)

**Comment by vanhauser-thc** [comment]
It is not fixed:
```
[!] cannot compile the test program or the custom mutator
```

**Comment by domenukk** [comment]
That would be a new bug. It worked well yesterday.
not sure if ignoring the test case is a good solution 🙃

**Comment by h1994st** [comment]
I think the new bug is due to the fail of compiling afl-clang-fast

**Pull request merged**